tox.ini

```diff
@@ -0,0 +1,144 @@
+# install tox from apt. Older than that, the user gets an upgrade warning.
+requires =
+    # renovate: datasource=pypi
+    tox==4.4.6
+    # renovate: datasource=pypi
+    tox-ignore-env-name-mismatch==0.2.0.post2
+# Allow tox to access the user's $TMPDIR environment variable if set.
+# This workaround is required to avoid circular dependencies for TMPDIR,
+# since tox will otherwise attempt to use the environment's TMPDIR variable.
+user_tmp_dir = {env:TMPDIR}
+
+[testenv]  # Default config for all environments. Overridable in each env.
+# We have many tests that create temporary files. Unless the user has set a
+# TMPDIR, this will prefer putting those temp files in $XDG_RUNTIME_DIR,
+# which will speed up those tests since they'll run on a ramdisk.
+env_tmp_dir = {user_tmp_dir:{work_dir}}/tox_tmp/{env_name}
+set_env =
+    SNAPCRAFT_IGNORE_YAML_BINDINGS = 1
+    TMPDIR={env_tmp_dir}
+    COVERAGE_FILE={env_tmp_dir}/.coverage_{env_name}
+
+[test]  # Base configuration for unit and integration tests
+deps = -r{tox_root}/requirements-devel.txt
+package = wheel
+allowlist_externals = mkdir
+commands_pre = mkdir -p results
+
+[testenv:test-{py38,py39,py310,py311,py312}]  # Configuration for all tests using pytest
+base = testenv, test
+description = Run unit tests with pytest
+labels =
+    py38, py310, py311: tests, unit-tests
+commands = pytest {tty:--color=yes} --cov --cov-report=xml:results/coverage-{env_name}.xml --junit-xml=results/test-results-{env_name}.xml {posargs:tests/unit}
+
+[testenv:test-legacy-{py38,py39,py310,py311,py312}]
+base = testenv, test
+description = Run legacy tests with pytest
+labels =
+    py38, py310, py311: tests, integration-tests
+commands = pytest {tty:--color=yes} --junit-xml=results/test-results-{env_name}.xml {posargs:tests/legacy}
+
+[testenv:test-noreq]
+base = testenv
+description = Run all tests without using requirements.txt
+package = wheel
+deps = python-apt@http://archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_2.4.0ubuntu1.tar.xz
+extras = dev
+allowlist_externals = mkdir
+commands_pre = mkdir -p results
+commands =
+    pytest {tty:--color=yes} --cov --cov-report=xml:results/coverage-{env_name}.xml --junit-xml=results/test-results-{env_name}.xml tests/unit {posargs}
+    pytest {tty:--color=yes} --junit-xml=results/legacy-test-results-{env_name}.xml tests/legacy {posargs}
+
+[lint]  # Standard linting configuration
+skip_install = true
+deps =
+    # renovate: datasource=pypi
+    black==23.1.0
+    # renovate: datasource=pypi
+    ruff==0.0.220
+    # renovate: datasource=pypi
+    codespell[tomli]==2.2.2
+    # renovate: datasource=pypi
+    yamllint==1.29.0
+    # renovate: datasource=pypi
+    isort==5.11.4
+env_dir = {work_dir}/linting
+runner = ignore_env_name_mismatch
+
+[shellcheck]
+find = git ls-files
+filter = file --mime-type -Nnf- | grep shellscript | cut -f1 -d:
+
+[testenv:lint-{black,ruff,shellcheck,codespell,yaml}]
```

Comment:
Is there a tox command that I can use to replace `make lint`?

Reply:
All of these have the `lint` label, so you can run `tox run -m lint` to run all the linters. (`tox lint -m lint` will show you which environments get run on that label.) It doesn't look like tox can list labels or factors though, so there's no way to discover this other than looking at the labels in the tox config.

The current labels in the tox config are:

* lint
* type (which is a subset of lint)
* format (runs black, ruff and codespell with auto-fixing)
* unit-tests (runs pytest for tests/unit in Python 3.8, 3.10 and 3.11)
* integration-tests (tests/legacy like above)
* tests (union of unit-tests and integration-tests)

Reply:
...I'm gonna document that in starbase

Reply:
Ah, `tox run -m lint`! Thanks!